tests/app/main/test_suppliers.py

```diff
@@ -2282,3 +2284,152 @@ def test_existing_org_size_sets_current_selection(self, data_api_client, existin
             doc = html.fromstring(res.get_data(as_text=True))
             selected_value = doc.xpath('//input[@name="organisation_size" and @checked="checked"]/@value')
             assert selected_value == expected_selection, 'The organisation size has not pre-populated correctly.'
+
+
+class TestEditSupplierVatNumber(BaseApplicationTest):
+    def setup_method(self, method):
+        super().setup_method(method)
+        self.data_api_client_patch = mock.patch("app.main.views.suppliers.data_api_client", autospec=True)
+        self.data_api_client = self.data_api_client_patch.start()
+        self.data_api_client.get_supplier.return_value = get_supplier(vatNumber=None)
+        self.login()
+
+    def teardown_method(self, method):
+        super().setup_method(method)
+        self.data_api_client_patch.stop()
+
+    def test_loads_page_for_supplier_with_no_vat_number(self):
+        with self.app.test_client():
+            res = self.client.get("/suppliers/vat-number/edit")
+            doc = html.fromstring(res.get_data(as_text=True))
+
+            assert res.status_code == 200
+            assert doc.xpath("//h1[normalize-space(string())='Are you registered for VAT?']")
+
+    @pytest.mark.parametrize('method, status_code', (('get', 200), ('post', 400)))
+    def test_get_and_post_load_already_completed_page_and_do_not_update_supplier_with_vat_number(
+        self, method, status_code
+    ):
+        with self.app.test_client():
+            self.data_api_client.get_supplier.return_value = get_supplier()
+
+            res = getattr(self.client, method)("/suppliers/vat-number/edit")
+            doc = html.fromstring(res.get_data(as_text=True))
+
+            assert res.status_code == status_code
+            assert doc.xpath("//h1[normalize-space(string())='Change your vat number']")
+            assert not self.data_api_client.update_supplier.called
+
+    def test_posting_valid_vat_code_updates_supplier(self):
+        with self.app.test_client():
+            res = self.client.post(
+                "/suppliers/vat-number/edit",
+                data={
+                    "vat_registered": "Yes",
+                    "vat_number": "GB123456789",
```

Comment:
Since you made this beautiful regex checking for several types of VAT numbers, both with and without prefixes, would it be worth it to parametrise this test and check that they all work? :)

Reply:
Never mind, I see you did it below, well done!